File: CustomAdapterDemo/app/src/main/java/com/example/customadapterdemo/UserAdapter.java
package com.example.customadapterdemo;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;


import androidx.annotation.NonNull;

import java.util.List;

public class UserAdapter extends ArrayAdapter<User> {

    private final Context context;
    private final List<User> userList;

    public UserAdapter(Context context, List<User> list) {
        super(context, 0, list);
        this.context = context;
        this.userList = list;
    }

    @SuppressLint("SetTextI18n")
    @NonNull
    @Override
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        View listItemView = convertView;
        if (listItemView == null) {
            listItemView = LayoutInflater.from(context).inflate(
                    R.layout.list_item_layout, parent, false); // Inflate layout
        }

        User currentUser= userList.get(position);

        ImageView userIcon = listItemView.findViewById(R.id.user_icon);
        TextView nameText = listItemView.findViewById(R.id.name_text);
        TextView cityText = listItemView.findViewById(R.id.city_text);

        nameText.setText(currentUser.getName());
        cityText.setText("from " + currentUser.getCity());

        return listItemView;
    }
}

File: CustomAdapterDemo1/app/src/main/java/com/example/customadapterdemo1/UserAdapter.kt
package com.example.customadapterdemo1
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView

class UsersAdapter(context: Context, users: ArrayList<User>) : ArrayAdapter<User>(context, 0, users) {

    private class ViewHolder {
        lateinit var name: TextView
        lateinit var home: TextView
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var view = convertView
        val viewHolder: ViewHolder

        if (view == null) {
            viewHolder = ViewHolder()
            view = LayoutInflater.from(context).inflate(R.layout.item_user, parent, false)
            viewHolder.name = view.findViewById(R.id.tvName)
            viewHolder.home = view.findViewById(R.id.tvHome)
            view.tag = viewHolder
        } else {
            viewHolder = view.tag as ViewHolder
        }

        val user = getItem(position)
        viewHolder.name.text = user?.name
        viewHolder.home.text = user?.hometown

        return view!!
    }
}